src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,196 @@
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        size++;
+        final Node<T> l = tail;
+        final Node<T> newNode = new Node<>(l, value, null);
+        tail = newNode;
+        if (l == null) {
+            head = newNode;
+        } else {
+            l.next = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        size++;
+        checkPositionIndex(index);
+
+        Node<T> newNode = defaultNode(value);
+
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size - 1) {
+            node(size - 2).next = newNode;
+            newNode.prev = node(size - 2);
+            tail = newNode;
+        } else {
+            Node<T> prevNode = node(index - 1);
+            newNode.next = prevNode.next;
+            newNode.prev = prevNode;
+            prevNode.next.prev = newNode;
+            prevNode.next = newNode;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        Node<T> pred;
+        Node<T> succ;
+        succ = null;
+        pred = tail;
+
+        for (T o : list) {
+            Node<T> newNode = new Node<>(pred, o, null);
+            if (pred == null) {
+                head = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            pred = newNode;
+        }
+
+        if (succ == null) {
+            tail = pred;
+        } else {
+            pred.next = succ;
+            succ.prev = pred;
+        }
+
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> x;
+        T oldVal;
+        if (index == size) {
+            x = tail;
+            x.value = value;
+            oldVal = x.value;
+            return oldVal;
+        }
+        x = node(index);
+        oldVal = x.value;
+        x.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
```

Comment:
combine these 2 loops and its if condition into one, because they have the same body